src/main/preload/externalAPI.ts

```diff
@@ -0,0 +1,459 @@
+    onOpenScreenShareModal: (listener) => ipcRenderer.on(DESKTOP_SOURCES_MODAL_REQUEST, () => listener()),
+    shareScreen: (sourceID, withAudio) => ipcRenderer.send(CALLS_WIDGET_SHARE_SCREEN, sourceID, withAudio),
+    joinCall: (opts) => ipcRenderer.invoke(CALLS_JOIN_CALL, opts),
+    sendJoinCallRequest: (callId) => ipcRenderer.send(CALLS_JOIN_REQUEST, callId),
+    onCallsError: (listener) => ipcRenderer.on(CALLS_ERROR, (_, err, callID, errMsg) => listener(err, callID, errMsg)),
+
+    // Utility
+    unregister: (channel) => ipcRenderer.removeAllListeners(channel),
+} as DesktopAPI);
+
+// Specific info for the testing environment
+if (process.env.NODE_ENV === 'test') {
+    contextBridge.exposeInMainWorld('testHelper', {
+        getViewInfoForTest: () => ipcRenderer.invoke(GET_VIEW_INFO_FOR_TEST),
+    });
+}
+
+/****************************************************************************
+ * window/document listeners
+ * These are here to perform specific tasks when global window or document events happen
+ * Avoid using these unless absolutely necessary
+ ****************************************************************************
+ */
+
+// Let the main process know when the window has finished resizing
+// This is to reduce the amount of white box that happens when expand the BrowserView
+window.addEventListener('resize', () => {
+    ipcRenderer.send(VIEW_FINISHED_RESIZING);
+});
+
+// Enable secure input on macOS clients when the user is on a password input
+let isPasswordBox = false;
+const shouldSecureInput = (element: {tagName?: string; type?: string} | null, force = false) => {
+    const targetIsPasswordBox = (element && element.tagName === 'INPUT' && element.type === 'password');
+    if (targetIsPasswordBox && (!isPasswordBox || force)) {
+        ipcRenderer.send(TOGGLE_SECURE_INPUT, true);
+    } else if (!targetIsPasswordBox && (isPasswordBox || force)) {
+        ipcRenderer.send(TOGGLE_SECURE_INPUT, false);
+    }
+
+    isPasswordBox = Boolean(targetIsPasswordBox);
+};
+window.addEventListener('focusin', (event) => {
+    shouldSecureInput(event.target as Element);
+});
+window.addEventListener('focus', () => {
+    shouldSecureInput(document.activeElement, true);
+});
+
+// exit fullscreen embedded elements like youtube - https://mattermost.atlassian.net/browse/MM-19226
+ipcRenderer.on('exit-fullscreen', () => {
+    if (document.fullscreenElement && document.fullscreenElement.nodeName.toLowerCase() === 'iframe') {
+        document.exitFullscreen();
+    }
+});
+
+// mattermost-webapp is SPA. So cache is not cleared due to no navigation.
+// We needed to manually clear cache to free memory in long-term-use.
+// http://seenaburns.com/debugging-electron-memory-usage/
+const CLEAR_CACHE_INTERVAL = 6 * 60 * 60 * 1000; // 6 hours
+setInterval(() => {
+    webFrame.clearCache();
+}, CLEAR_CACHE_INTERVAL);
+
+/****************************************************************************
```

Comment:
You can't actually load two separate preload files, you only get one.

I suppose we could have just had a central module loading both, and maybe some logic as to which to load based on which Web App client was running, but there were too many variables and the complexity was higher.

We still need to load both for the time being, but we will be removing the legacy code eventually.

Reply:
I was thinking that we have one `preload/index.js` file that we preload which just imports `preload/newapi.js` and `preload/oldapi.js` or something similar. Then, we could just remove the legacy code by deleting the old one

Reply:
That's the only thing I'd draw attention to. It's worth looking over/testing the legacy code since I reformatted it a bit, but fundamentally it works the same.

Reply:
Okay, I did that previously, so the rest of that looks good to me then